src/AstTransforms.cpp

```diff
@@ -221,6 +222,117 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
     return changed;
 }
 
+std::string MaterializeSingletonAggregationTransformer::findUniqueVariableName(const AstClause& clause) {
+    int counter = 0;
+    std::set<std::string> variableNames;
+    visitDepthFirst(clause, [&](const AstVariable& variable) { variableNames.insert(variable.getName()); });
+    std::string candidateVariableName = "z";  // completely arbitrary
+    while (variableNames.find(candidateVariableName) != variableNames.end()) {
+        candidateVariableName = "z" + toString(counter++);
+    }
+    return candidateVariableName;
+}
+
+std::string MaterializeSingletonAggregationTransformer::findUniqueAggregateRelationName(
+        const AstProgram& program) {
+    int counter = 0;
+    auto candidate = "__agg_rel_" + toString(counter++);
+    while (getRelation(program, candidate) != nullptr) {
+        candidate = "__agg_rel_" + toString(counter++);
+    }
+    return candidate;
+}
+
+bool MaterializeSingletonAggregationTransformer::transform(AstTranslationUnit& translationUnit) {
+    AstProgram& program = *translationUnit.getProgram();
+    std::set<std::pair<AstAggregator*, AstClause*>> pairs;
+    // collect references to clause / aggregate pairs
+    // let's just make it for one literal for now. Following Martin's approach is a better idea.
+    visitDepthFirst(program, [&](const AstClause& clause) {
+        visitDepthFirst(clause, [&](const AstAggregator& agg) {
+            if (!isSingleValued(agg)) {
+                return;
+            }
+            auto* foundAggregate = const_cast<AstAggregator*>(&agg);
+            auto* foundClause = const_cast<AstClause*>(&clause);
+            pairs.insert(std::make_pair(foundAggregate, foundClause));
+        });
+    });
+    for (auto pair : pairs) {
+        AstAggregator* aggregate = pair.first;
+        AstClause* clause = pair.second;
+        // synthesise an aggregate relation
+        // __agg_rel_0()
+        auto* aggRel = new AstRelation();
+        auto* aggHead = new AstAtom();
+        auto* aggClause = new AstClause();
+        std::string aggRelName = findUniqueAggregateRelationName(program);
+        aggRel->setQualifiedName(aggRelName);
+        aggHead->setQualifiedName(aggRelName);
+        // create a synthesised variable to replace the aggregate term!
+        std::string variableName = findUniqueVariableName(*clause);
+        auto* variable = new AstVariable(variableName);
+        // __agg_rel_0(z) :- ...
+        aggHead->addArgument(std::unique_ptr<AstArgument>(variable));
```

Comment:
This is moving `variable` into the AstAtom - so `variable` is no longer defined. You then use it again to make `equalityLiteral`.

Reply:
Oh... I thought if I wrap a raw pointer in the unique ptr then if the old raw pointer is still hanging around I can use it for other stuff. This is not possible? I need to clone before I let it go into the aggHead?

Reply:
I think what I said is wrong. `variable` is still fine, since aggHead still exists. There is a use after delete issue occurring with it, according to the memory sanitiser, but I'm not so sure where, now.

I ran the tests locally, with the memory sanitiser enabled (which is similar to valgrind, if you prefer).
I used
```
make distclean;./bootstrap && ./configure --enable-debug --enable-sanitise-memory && TESTSUITEFLAGS=-j16 SOUFFLE_CONFS=-j1 SOUFFLE_CATEGORY=Semantic make check -j16
```
You would want to remove the category/confs for more general testing.

Reply:
OHHH I see what you are saying. Once the pointer became wrapped up, maybe later the unique pointer falls out of scope and the variable gets deallocated already. This means maybe I should clone way before

Reply:
Okay I will try this too!

Reply:
Okay now I also understand why it can't be that too. ok

Reply:
> > who should upstream be? :( I need to set this right? Is upstream my repo or main souffle or your repo?
> 
> My workflow has origin as my repo and upstream as souffle-lang. If I do `git fetch --all` then all the links to remote stuff is updated. I can then do `git rebase upstream/master` to rebase onto that - similar to using `git merge`. If I do `git rebase -i ...` then I can edit each commit as I see fit.
> 
> This is all to make the history look neater. It's not essential.

Okay, I don't really get how to connect my fork to the original souffle-lang. How do I actually make it be aware that it exists at all? Do I need like another branch, and that one is souffle-lang? I kind of don't really get it, because upstream is a branch, and so if i set upstream to souffle-lang, doesn't it not make sense bcause souffle-lang has many branches?

Reply:
Try these git commands to illustrate:
```
git remote
```
This lists all your remote repositories  - if you've only cloned your own repo you will have one called 'origin'.

```
git remote add mmcgr https://github.com/mmcgr/souffle
```
That adds another repo called mmcgr.
```
git fetch mmcgr
```
Now we've got all the branch info from mmcgr
```
git branch -r
```
Lists all the branches in remote repositories

And an alias that you may still have installed from ages ago:
```
git lg
```
to show the git commit tree.

You can only push to a remote repository if you have access, which will generally just be your own repo.

So, now you can add souffle-lang as upstream with
```
git remote add upstream https://github.com/souffle-lang/souffle
git fetch --all
```

Reply:
Thank you for this!